### ISSUE TITLE
If there is an upstream result, I would like to get it....

### ISSUE THREAD

**Issue by alcidesv** [snoyberg/conduit#199]
It would be nice to have the upstream result of a conduit, if such result is produced. Like fuseBoth but with a Maybe r1. More details below, taken literally from  [this question](http://stackoverflow.com/questions/28523958/haskell-conduit-is-it-possible-to-optionally-have-the-result-of-a-source/28526509#28526509) at StackOverflow. 

I have the following types:

   type Headers = [(ByteString, ByteString)] 
 
    type Footers = [(ByteString, ByteString)]
    
    type PrincipalStream = (Headers, PushedStreams, DataAndConclussion)

    type DataAndConclusion = ConduitM () ByteString IO Footers

These types represent very well a cancellable HTTP/2 stream: after the headers are produced, the body of the request can be fetched from the conduit. And then the footers can be sent, if the stream was not cancelled by the browser in the middle and all the ByteString's were fetched from the source. 

The idea of the second type being "produce a lot of ByteStrings, and *if* you can produce all of them, return a Footers". The consumer of DataAndConclusion may have no need to consume all its items, and in that case the return wouldn't be reached. Which is precisely the behavior that I need. But *when* the end of the source is reached, I would like to have the produced Footers. This would be useful for example if the DataAndConclusions were incrementally computing an MD5 and such MD5 was only needed if the entire message was processed by the downstream (for example, downstream could be simply sending it through the network, but it doesn't make sense to finish computing and send the MD5 if the socket was closed before the last piece was sent by downstream). This is the kind of behavior suggested by the HTTP/2 standard for the footers. 

So, basically I want to have something with this signature to consume a DataAndConclusions:

     type MySink = Sink ByteString IO ()
   
     mySink :: MySink 
     mySink = ... 

     maybeUpstreamResult  :: ConduitM () a2 m r1 -> ConduitM a2 Void m r2 -> m (Maybe r1)

Can we have something like maybeUpstreamResult?

**Comment by snoyberg** [comment]
It's a good idea. I've just pushed a commit that adds a `fuseBothMaybe`. Can you give it a shot and let me know if it solves your use case?

**Comment by alcidesv** [comment]
Thanks a lot! I checked the new feature, it works perfectly!
